src/goto-instrument/synthesizer/loop_invariant_synthesizer_base.h

```diff
@@ -0,0 +1,51 @@
+/*******************************************************************\
+
+Module: Loop Invariant Synthesizer Interface
+
+Author: Qinheping Hu
+
+\*******************************************************************/
+
+/// \file
+/// Loop Invariant Synthesizer Interface
+
+#ifndef CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_LOOP_INVARIANT_SYNTHESIZER_BASE_H
+#define CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_LOOP_INVARIANT_SYNTHESIZER_BASE_H
+
+#include <goto-programs/goto_model.h>
+
+#include "synthesizer_utils.h"
+
+#define OPT_SYNTHESIZE_LOOP_INVARIANTS "(synthesize-loop-invariants)"
+#define HELP_LOOP_INVARIANT_SYNTHESIZER                                        \
+  " --synthesize-loop-invariants\n"                                            \
+  "                              synthesize and apply loop invariants\n"
+
+class messaget;
+
+/// A base class for loop invariant synthesizers.
+/// Provides a method for synthesizing loop invariants in a given `goto_model`.
+///
+/// Derived class should clarify what types of `goto_model` they targets, e.g.,
+/// a `goto_model` contains only memory safety checks, or a `goto_model`
+/// contains both memory safety checks and correctness checks.
+class loop_invariant_synthesizer_baset
+{
+public:
+  loop_invariant_synthesizer_baset(const goto_modelt &goto_model, messaget &log)
+    : goto_model(goto_model), log(log)
+  {
+  }
+  virtual ~loop_invariant_synthesizer_baset() = default;
+
+  /// Synthesize loop invariants with which all checks in `goto_model`
+  /// succeed. The result is a map from `loop_idt` ids of loops to `exprt`
+  /// the goto-expression representation of synthesized invariants.
+  virtual invariant_mapt synthesize() = 0;
```

Comment:
I realise that this is pretty late on for requesting architectural changes but would it be possible for this to be:

```
virtual exprt synthesize(loop_idt);
```

or even

```
virtual exprt synthesize(irep_idt function_id, unsigned int loop_number);
```

The existing functionality can be kept as `synthesize_all` of `generate_invariant_mapt`.

The reason I ask is that then the component can be used in other tools.  If the input program is large and the invariant generation technique is expensive (many of the fun ones are) then having to generate all of them up front is really not great.  One obvious thing to do is to try to use cheap techniques first and then use heavier ones on just the loops that are not currently proven.  Being able to generate invariants for one loop at a time will allow this and other incremental approaches.

Reply:
I agree that finer-grain API is a good idea and added it in the latest commit. 
My only concern is how to decide whether a loop invariant is good enough when we synthesize it for only one specified loop. In the `synthesize_all` function, we terminate and return the synthesized loop invariants when all checks in the goto-model pass. But if we are targeting only one loop, do we want to synthesized the loop invariant that pass all checks dominated by the specified loops? Or do we just do the best-effort synthesize to generate a loop invariant that pass checks as many as possible? 
I will think about it more carefully and specify the purpose of `synthesize(loop_idt)` in the later PR. Suggestions are welcome.

Reply:
A number of overlapping thoughts:

* As a general rule, the more incremental and lazy /https://vm.tiktok.com/ZMNGcDcR7/ on demand you can be the better your code will scale and the more useful it will be.
* There is nothing to stop you implementing `synthesize(loop_idt)` as "do the global computation, cache the map and then return the relevant entry".
* It seems that perhaps I had misunderstood what your goals are.  I think there are maybe two different scales of functionality that this could implement.  A. (my original, probably mistaken, understanding) this is a way of getting invariants that /might/ be useful for others with the main control loop in the caller or B. this is full, whole program verification algorithm that returns the invariants needed to verify the whole program, main control loop is in the callee.
* In the case of A, other implementations, such as connecting to the abstract interpreter or implementing the Daikon algorithm or similar could be used.  These have different stopping points.  For B these might not make as much sense.
* In the case of B. maybe `goto-checker` is a better interface or whatever @tautschnig uses to generate invariant-based proofs for SV-COMP.
* Maybe the API should say what can be expected of the invariants returned.  Presumably they are loop invariants (even if the loop contains an assertion which is false?).  Is their base-case from the whole program or just local? Are they guaranteed to be non-trivial (not implied by the existing invariant(s)), or not "obviously" trivial?  Are they guaranteed to prove any assertions that are true that are in the loop?  Do they have wider properties?

Depending on which of these are required, I could probably easily enough come up with a second implementation of the API using the abstract interpreter.

Reply:
Thank you!
Yes, my original goals of this `enumerative_loop_invariant_synthesizert` were the scale B: synthesis with guarantees/hard constraints. You are right that there can be varied guarantees other than only B---all checks success.
So I think for the base class, we only expect the returned `exprt` are _loop invariants_ in the sense that
1. the instrumented loop invariant checks success no matter other assertions fail or success---so we just ignore even if an assertion is false in the loop;
3. the base-case is from the whole program with a main function as entry.
Derived classes can have their own guarantees. Each derived class should specify what the users can expect about the returned loop invariants. And users can choose which synthesizer to use with options under the `--synthesize-loop-invariants` flag.

Reply:
Sounds like a fair resolution.  Would you mind creating a PR to document this semantics?

Reply:
I created a [PR#7089](https://github.com/diffblue/cbmc/pull/7089) to document this semantics.

Reply:
Thank you.